packages/client/src/client.ts

```diff
@@ -349,4 +370,67 @@ export class DefaultDelegatedRoutingV1HttpApiClient implements DelegatedRoutingV
       }
     }
   }
+
+  //
+  // and caches GET requests
+  /**
+   * makeRequest has two features:
+   * - Ensures only one concurrent request is made for the same URL
+   * - Caches GET requests if the Cache API is available
+   */
+  async #makeRequest (url: string, options: RequestInit): Promise<Response> {
+    const requestMethod = options.method ?? 'GET'
+    const key = `${requestMethod}-${url}`
+
+    // Only try to use cache for GET requests
+    if (requestMethod === 'GET') {
+      const cachedResponse = await this.cache?.match(url)
+      if (cachedResponse != null) {
+        // Check if the cached response has expired
+        const expires = parseInt(cachedResponse.headers.get('x-cache-expires') ?? '0', 10)
+        if (expires > Date.now()) {
+          log('returning cached response for %s', key)
+          return cachedResponse
+        } else {
+          // Remove expired response from cache
+          await this.cache?.delete(url)
+        }
+      }
+    }
+
+    // Check if there's already an in-flight request for this URL
+    const existingRequest = this.inFlightRequests.get(key)
+    if (existingRequest != null) {
+      const response = await existingRequest
+      log('deduplicating outgoing request for %s', key)
+      return response.clone()
+    }
+
+    // Create new request and track it
+    const requestPromise = fetch(url, options).then(async response => {
+      // Only cache successful GET requests
+      if (this.cache != null && response.ok && requestMethod === 'GET') {
+        const expires = Date.now() + this.cacheTTL
+        const headers = new Headers(response.headers)
+        headers.set('x-cache-expires', expires.toString())
+
+        // Create a new response with expiration header
+        const cachedResponse = new Response(response.clone().body, {
+          status: response.status,
+          statusText: response.statusText,
+          headers
+        })
+
+        await this.cache.put(url, cachedResponse)
+      }
+      return response
+    }).finally(() => {
+      // Clean up the tracked request when it completes
+      this.inFlightRequests.delete(key)
+    })
+
+    this.inFlightRequests.set(key, requestPromise)
+    const response = await requestPromise
+    return response.clone()
```

Comment:
Do we need to clone the original response? From looking at the code it doesn't seem like we do.

Reply:
Probably not. I think there was a reason that no longer holds.